docs/upgrade-to-v3.md

```diff
@@ -0,0 +1,244 @@
+1. Replace any reference to `com.microsoft.graph.
+
+> Note: if your project implements its own authentication provider, you might want to review the list of providers available with Azure identity to review whether a custom provider is still required. If a custom provider is still required, you should extend `BaseAuthenticationProvider` which provides a method to check whether a spefic request should be authenticated based on it's URL.
+
+### Modern date and time API
+
+To provide a modern, more reliable, and faster date and time API, `Calendar` properties have been replaced by `OffsetDateTime` properties:
+
+1. Replace any `Calendar` assignment by `OffsetDateTime`.
+
+### Fluent method parameter sets and support for optional parameters
+
+To provide support for optional parameters for OData methods (actions, functions), the SDK now makes use of parameter classes with fluent builders:
+
+1. Replace any OData method reference by it's parameter set equivalent.
+
+   Example:
+
+   ```Java
+   final WorkbookRange workbookRange = graphClient.me()
+             .drive()
+             .items("{id}")
+             .workbook()
+             .worksheets("{id|name}")
+             .range("A1:B2");
+   ```
+
+   Now becomes:
+
+   ```Java
+   final WorkbookRange workbookRange = graphClient.me()
+             .drive()
+             .items("{id}")
+             .workbook()
+             .worksheets("{id|name}")
+             .range(WorkbookWorksheetRangeParameterSet
+               .newBuilder()
+               .withAddress("A1:B2")
+               .build());
+   ```
+
+### Futures vs Callbacks
+
+To provide a standard and modern way of executing requests asynchronuously to SDK users, the custom defined `ICallback` interface has been removed and the API now returns `CompletableFuture`.
+
+1. Replace any of the code built the following way
+
+    ```Java
+    graphClient
+    .me()
+    .drive()
+    .buildRequest()
+    .get(new ICallback<Drive>() {
+      @Override
+      public void success(final Drive result) {
+        System.out.println("Found Drive " + result.id);
+      }
+    });
+    ```
+
+    By:
+
+    ```Java
+    graphClient
+    .me()
+    .drive()
+    .buildRequest()
+    .getAsync()
+    .thenApply(result -> {
+        System.out.println("Found Drive " + result.id);
+    });
+    ```
+
+### Non-generated code moved to the core library
+
+The service library (microsoft-graph) was designed to contain only code that has been generated automatically from the API metadata. All code that has been manually handcrafted should live in the core library (microsoft-graph-core). This principle makes reusability of code easier for consumers and it was not respected until this new version of the SDK. All the code living under anything but `com.microsoft.graph.models` or `.requests` have been moved to the same package name in the core asset. This move should be transparent to consummers expect for the following that needs to be updated:
```

Comment:
We have enough context to deduce the full namespace, but it may not be that straightforward for someone occasionally using the SDK.
```suggestion
The service library (microsoft-graph) was designed to contain only code that has been generated automatically from the API metadata. All code that has been manually handcrafted should live in the core library (microsoft-graph-core). This principle makes reusability of code easier for consumers and it was not respected until this new version of the SDK. All the code living under anything but `com.microsoft.graph.models` or `com.microsoft.graph.requests` have been moved to the same package name in the core asset. This move should be transparent to consumers, except for the following that needs to be updated:
```